pkg/core/bundle_list.go

```diff
@@ -25,29 +28,339 @@ func minInt(a, b int) int {
+	var (
+		wg, wg2 sync.WaitGroup
+	)
+
+	settings := defaultCoreSettings()
 	for _, bApply := range opts {
 		bApply(&settings)
 	}
 
-	// Get a list
-	e := RepoExists(repo, store)
-	if e != nil {
-		return nil, e
+	batchChan := make(chan model.BundleDescriptors)
+	errorChan := make(chan error, 1) // buffering early errors avoids blocking without dropping the message
+
+	if err := RepoExists(repo, store); err != nil {
+		errorChan <- err
+		close(batchChan)
+		close(errorChan)
+		return batchChan, errorChan, &wg
 	}
 
-	// this call is not made async yet
-	ks, _, err := store.KeysPrefix(context.Background(), "", model.GetArchivePathPrefixToBundles(repo), "/", maxMetaFilesToProcess)
-	if err != nil {
-		return nil, err
+	// signaling channels
+	doneWithKeysChan := make(chan struct{}, 1)
+	doneWithBundlesChan := make(chan struct{}, 1)
+
+	if settings.doneChannel != nil {
+		// watch for an interruption signal requested by caller
+		wg.Add(1)
+		go watchForInterrupts(settings.doneChannel, &wg, doneWithKeysChan, doneWithBundlesChan)
 	}
 
+	keysChan := make(chan []string)
+	keyErrorChan := make(chan error)
+	bundleErrorChan := make(chan error)
+
+	// listening to errors from slave goroutines
+	wg.Add(1)
+	go watchForErrors(keyErrorChan, bundleErrorChan, errorChan, doneWithKeysChan, doneWithBundlesChan, &wg)
+
+	// starting keys retrieval
+	wg.Add(1)
+	go fetchKeys(repo, store, settings, keysChan, keyErrorChan, doneWithKeysChan, &wg) // scan for key batches
+
+	// start bundle metadata retrieval
+	wg.Add(1)
+	go fetchBundles(repo, store, settings, keysChan, batchChan, bundleErrorChan, doneWithBundlesChan, &wg)
+
+	// cleanup internal signaling channels
+	wg2.Add(1)
+	go func(wg, wg2 *sync.WaitGroup, inputChans ...chan<- struct{}) {
+		defer wg2.Done()
+		wg.Wait()
+		for _, ch := range inputChans {
+			close(ch)
+		}
+	}(&wg, &wg2, doneWithKeysChan, doneWithBundlesChan)
+
+	// return at once. Caller may chose to wait on returned WaitGroup
+	return batchChan, errorChan, &wg2
+}
+
+// watchForInterrupts broadcasts a done signal to several output channels
+func watchForInterrupts(doneChan <-chan struct{}, wg *sync.WaitGroup, outputChans ...chan<- struct{}) {
+	defer func() {
+		wg.Done()
+	}()
+
+	if _, interrupt := <-doneChan; interrupt {
+		for _, outputChan := range outputChans {
+			outputChan <- struct{}{}
+		}
+	}
+}
+
+// watchForErrors listens on 2 independent errors channels and report back to a main error channel.
+//
+// NOTE: this one does not need signaling, since any interruption bubbles up as error.
+func watchForErrors(
+	errorChan1, errorChan2 <-chan error,
+	outputChan chan<- error,
+	doneChan1, doneChan2 chan<- struct{}, wg *sync.WaitGroup) {
+	defer func() {
+		close(outputChan)
+		wg.Done()
+	}()
+
+	var (
+		err                            error
+		errorChan1Open, errorChan2Open bool = true, true
+	)
+	for errorChan1Open || errorChan2Open { // wait on both channels to close
+		select {
```

Comment:
[recall](https://golang.org/ref/spec#Select_statements)

> If one or more of the communications can proceed, a single one that can proceed is chosen via a uniform pseudo-random selection

so to prevent the enclosing `for` loop from being stochastic, i suggest adding a `switch` to read from precisely one or the other channel once one of the two is closed.